torchtitan/train.py

```diff
@@ -414,12 +414,20 @@ def forward_backward_step(
 
         inputs = input_dict["input"]
         extra_inputs = {k: v for k, v in input_dict.items() if k != "input"}
-        # Create the FlexAttention mask according to the input
-        if getattr(self.model_args, "use_flex_attn", False):
-            cp_mesh = (
-                parallel_dims.world_mesh["cp"] if parallel_dims.cp_enabled else None
-            )
-            init_attention_mask(inputs, self.tokenizer.eos_id, cp_mesh)
+
+        # Create Attention masks. Currently, if SDPA is used, the masks should be None.
+        # However, this may change in the future if SDPA supports varlen.
+        use_flex_attn = getattr(self.model_args, "use_flex_attn", False)
+        cp_mesh = parallel_dims.world_mesh["cp"] if parallel_dims.cp_enabled else None
+        create_mask_fn = get_create_mask_fn(
+            use_flex_attn=use_flex_attn,
+            cp_mesh=cp_mesh,
+        )
+        attention_masks = model_parts[0].get_attention_masks(
+            create_mask_fn=create_mask_fn,
+            batch=inputs,
+            eos_id=self.tokenizer.eos_id,
```

Comment:
Can we make this more general for other models such as vlm?
For example, we might need `pixel_values` (and `grid_thw` for mask) in `extra_inputs`, as well as some other special tokens like `img_id, soi_id, eoi_id`. 
Locally, sometimes we need to pass in a bunch more special tokens for a specific model/mask type, so pass in the entire tokenizer (or a dict/dataclass of special tokens) would be much more ergonomic.

```suggestion
        attention_masks = model_parts[0].get_attention_masks(
            create_mask_fn=create_mask_fn,
            batch=inputs,
            tokenizer=self.tokenizer,
            **extra_inputs,
```

Reply:
Nice suggestion. I was wondering other models as I didn't have good experiments on VLM. I'll definitely take a look.